File: sockets/java/src/ServidorChat.java
import java.io.*;
import java.net.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Servidor de chat que permite comunicación entre múltiples usuarios
 */
public class ServidorChat {
    private static final int PUERTO = 8082;
    private static final AtomicInteger contadorClientes = new AtomicInteger(0);
    private static final Map<String, PrintWriter> clientes = new ConcurrentHashMap<>();
    
    public static void main(String[] args) {
        System.out.println("=== SERVIDOR DE CHAT MULTI-USUARIO ===");
        System.out.println("Iniciando servidor de chat en puerto " + PUERTO);
        
        try (ServerSocket servidor = new ServerSocket(PUERTO)) {
            System.out.println("Servidor de chat iniciado. Esperando usuarios...");
            
            while (true) {
                Socket cliente = servidor.accept();
                int numeroCliente = contadorClientes.incrementAndGet();
                
                System.out.println("Nuevo usuario conectado desde: " + 
                    cliente.getInetAddress());
                
                Thread hiloCliente = new Thread(
                    new ManejadorUsuario(cliente, numeroCliente));
                hiloCliente.start();
            }
            
        } catch (IOException e) {
            System.err.println("Error del servidor de chat: " + e.getMessage());
        }
    }
    
    /**
     * Difunde un mensaje a todos los usuarios conectados
     */
    public static synchronized void difundirMensaje(String mensaje, String remitente) {
        String timestamp = LocalDateTime.now().format(
            DateTimeFormatter.ofPattern("HH:mm:ss"));
        String mensajeCompleto = "[" + timestamp + "] " + mensaje;
        
        System.out.println("Difundiendo: " + mensajeCompleto);
        
        // Crear una copia para evitar ConcurrentModificationException
        Map<String, PrintWriter> clientesCopia = new HashMap<>(clientes);
        
        for (Map.Entry<String, PrintWriter> entry : clientesCopia.entrySet()) {
            if (!entry.getKey().equals(remitente)) {
                try {
                    entry.getValue().println(mensajeCompleto);
                } catch (Exception e) {
                    System.err.println("Error enviando mensaje a " + entry.getKey());
                    clientes.remove(entry.getKey());
                }
            }
        }
    }
    
    /**
     * Clase que maneja un usuario del chat
     */
    static class ManejadorUsuario implements Runnable {
        private final Socket socket;
        private final int numeroUsuario;
        private String nombreUsuario;
        
        public ManejadorUsuario(Socket socket, int numeroUsuario) {
            this.socket = socket;
            this.numeroUsuario = numeroUsuario;
        }
        
        @Override
        public void run() {
            try (
                BufferedReader entrada = new BufferedReader(
                    new InputStreamReader(socket.getInputStream()));
                PrintWriter salida = new PrintWriter(
                    socket.getOutputStream(), true)
            ) {
                // Solicitar nombre de usuario
                salida.println("Bienvenido al chat! ¿Cuál es tu nombre?");
                nombreUsuario = entrada.readLine();
                
                if (nombreUsuario == null || nombreUsuario.trim().isEmpty()) {
                    nombreUsuario = "Usuario" + numeroUsuario;
                }
                
                // Registrar usuario
                clientes.put(nombreUsuario, salida);
                difundirMensaje(nombreUsuario + " se unió al chat", nombreUsuario);
                
                salida.println("Conectado como: " + nombreUsuario);
                salida.println("Usuarios conectados: " + clientes.size());
                salida.println("Escribe '/salir' para desconectar o '/usuarios' para ver usuarios");
                
                String mensaje;
                while ((mensaje = entrada.readLine()) != null) {
                    if ("/salir".equalsIgnoreCase(mensaje.trim())) {
                        break;
                    } else if ("/usuarios".equalsIgnoreCase(mensaje.trim())) {
                        salida.println("Usuarios conectados: " + 
                            String.join(", ", clientes.keySet()));
                    } else {
                        difundirMensaje(nombreUsuario + ": " + mensaje, nombreUsuario);
                    }
                }
                
            } catch (IOException e) {
                System.err.println("Error con usuario " + nombreUsuario + ": " + e.getMessage());
            } finally {
                // Desconectar usuario
                if (nombreUsuario != null) {
                    clientes.remove(nombreUsuario);
                    difundirMensaje(nombreUsuario + " salió del chat", nombreUsuario);
                    System.out.println("Usuario " + nombreUsuario + " desconectado");
                }
                
                try {
                    socket.close();
                } catch (IOException e) {
                    System.err.println("Error cerrando socket: " + e.getMessage());
                }
            }
        }
    }
}

File: sockets/python/servidor_chat.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Servidor de chat multi-usuario que permite comunicación entre clientes
Cada cliente se maneja en un hilo separado y los mensajes se difunden a todos
"""

import socket
import threading
import datetime

HOST = 'localhost'
PUERTO = 8082

# Diccionario thread-safe para almacenar clientes conectados
clientes_conectados = {}
lock = threading.Lock()

def main():
    print("=== SERVIDOR DE CHAT ===")
    print(f"Iniciando servidor de chat en {HOST}:{PUERTO}")
    
    # Crear socket del servidor
    servidor = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    try:
        # Permitir reutilizar la dirección
        servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        servidor.bind((HOST, PUERTO))
        servidor.listen(10)
        
        print("Servidor de chat iniciado. Esperando usuarios...")
        
        while True:
            # Aceptar conexión del cliente
            cliente_socket, direccion_cliente = servidor.accept()
            
            # Crear hilo para manejar este cliente
            hilo_cliente = threading.Thread(
                target=manejar_cliente_chat,
                args=(cliente_socket, direccion_cliente)
            )
            hilo_cliente.daemon = True
            hilo_cliente.start()
            
    except KeyboardInterrupt:
        print("\nServidor de chat interrumpido")
    except Exception as e:
        print(f"Error del servidor: {e}")
    finally:
        servidor.close()
        print("Servidor de chat cerrado")

def manejar_cliente_chat(cliente_socket, direccion_cliente):
    """
    Maneja un cliente del chat en un hilo separado
    """
    nombre_cliente = None
    
    try:
        # Pedir nombre del usuario
        cliente_socket.send("Ingresa tu nombre: ".encode('utf-8'))
        nombre_cliente = cliente_socket.recv(1024).decode('utf-8').strip()
        
        if not nombre_cliente:
            nombre_cliente = f"Usuario_{direccion_cliente[1]}"
        
        # Agregar cliente al diccionario (thread-safe)
        with lock:
            clientes_conectados[nombre_cliente] = cliente_socket
            numero_clientes = len(clientes_conectados)
        
        print(f"Usuario '{nombre_cliente}' conectado desde {direccion_cliente}")
        
        # Notificar a todos que se unió un nuevo usuario
        mensaje_union = f"*** {nombre_cliente} se unió al chat ({numero_clientes} usuarios conectados) ***"
        difundir_mensaje(mensaje_union, excluir=nombre_cliente)
        
        # Enviar mensaje de bienvenida al nuevo usuario
        timestamp = datetime.datetime.now().strftime("%H:%M:%S")
        bienvenida = (
            f"\n=== BIENVENIDO AL CHAT ===\n"
            f"Hora: {timestamp}\n"
            f"Usuarios conectados: {numero_clientes}\n"
            f"Comandos: /usuarios, /salir\n"
            f"==========================\n"
        )
        cliente_socket.send(bienvenida.encode('utf-8'))
        
        # Bucle principal del chat
        while True:
            datos = cliente_socket.recv(1024)
            if not datos:
                break
                
            mensaje = datos.decode('utf-8').strip()
            
            if mensaje.lower() == "/salir":
                break
            elif mensaje.lower() == "/usuarios":
                with lock:
                    lista_usuarios = list(clientes_conectados.keys())
                respuesta = f"Usuarios conectados ({len(lista_usuarios)}): {', '.join(lista_usuarios)}\n"
                cliente_socket.send(respuesta.encode('utf-8'))
            else:
                # Difundir mensaje a todos los usuarios
                timestamp = datetime.datetime.now().strftime("%H:%M:%S")
                mensaje_completo = f"[{timestamp}] {nombre_cliente}: {mensaje}"
                print(mensaje_completo)  # Log en el servidor
                difundir_mensaje(mensaje_completo, excluir=nombre_cliente)
                
    except Exception as e:
        print(f"Error con usuario '{nombre_cliente}': {e}")
    finally:
        # Remover cliente del diccionario
        if nombre_cliente:
            with lock:
                clientes_conectados.pop(nombre_cliente, None)
                numero_clientes = len(clientes_conectados)
            
            # Notificar a todos que el usuario se desconectó
            mensaje_salida = f"*** {nombre_cliente} abandonó el chat ({numero_clientes} usuarios restantes) ***"
            difundir_mensaje(mensaje_salida)
            print(f"Usuario '{nombre_cliente}' desconectado")
        
        cliente_socket.close()

def difundir_mensaje(mensaje, excluir=None):
    """
    Envía un mensaje a todos los clientes conectados
    """
    mensaje_con_newline = mensaje + "\n"
    clientes_desconectados = []
    
    with lock:
        for nombre, socket_cliente in clientes_conectados.items():
            if nombre != excluir:  # No enviar al remitente
                try:
                    socket_cliente.send(mensaje_con_newline.encode('utf-8'))
                except:
                    # Cliente desconectado, marcarlo para eliminación
                    clientes_desconectados.append(nombre)
        
        # Eliminar clientes desconectados
        for nombre in clientes_desconectados:
            clientes_conectados.pop(nombre, None)
            print(f"Cliente {nombre} eliminado (conexión perdida)")

if __name__ == "__main__":
    main()